src/v/net/conn_quota.cc

```diff
@@ -0,0 +1,440 @@
+            continue;
+        }
+
+        total_released += co_await container().invoke_on(
+          i, [addr, one_time](conn_quota& cq) {
+              return cq.reclaim_from(addr, one_time);
+          });
+    }
+
+    allowance.available = std::min(
+      allowance.max, allowance.available + total_released);
+}
+
+/**
+ * Called on the non-home shards for `addr` during a reclaim of
+ * borrowed tokens.  Returns the number of borrowed tokens, and
+ * sets the reclaim flag to true so that subsequently freed tokens
+ * will be dispatched back to the home shard.
+ *
+ * @param one_time if true, do not leave the shard in reclaim mode,
+ *                 just grab any borrowed units they have currently.
+ */
+uint32_t conn_quota::reclaim_from(ss::net::inet_address addr, bool one_time) {
+    vlog(rpc::rpclog.trace, "reclaim_from({})", addr);
+    if (addr == ss::net::inet_address()) {
+        total_remote.reclaim = !one_time;
+        return std::exchange(total_remote.borrowed, 0);
+    } else {
+        ip_remote[addr].reclaim = !one_time;
+        return std::exchange(ip_remote[addr].borrowed, 0);
+    }
+}
+
+/**
+ * Called on non-home shards for `addr` after a previous call
+ * to `reclaim_from`, to unset the reclaim flag and permit this
+ * non-home shard to store borrowed tokens again.
+ *
+ * This is invoked from the home shard when it deems that there
+ * is no longer pressure for tokens.
+ */
+void conn_quota::cancel_reclaim_from(ss::net::inet_address addr) {
+    vlog(rpc::rpclog.trace, "cancel_reclaim_from({})", addr);
+    if (addr == ss::net::inet_address()) {
+        total_remote.reclaim = false;
+    } else {
+        auto found = ip_remote.find(addr);
+        if (found != ip_remote.end()) {
+            found->second.reclaim = false;
+        }
+    }
+}
+
+/**
+ * I am the home shard for this address.  Acquire tokens, reclaiming
+ * if necessary.
+ */
+ss::future<bool> conn_quota::home_get_units(ss::net::inet_address addr) {
+    assert_on_home(addr);
+
+    auto& allowance = get_home_allowance(addr);
+    vlog(rpc::rpclog.trace, "home_get_units({}) allowance={}", addr, allowance);
+
+    bool result = try_get_units(allowance);
+    if (result || allowance.reclaim) {
+        // If we got a token, or we didn't and are already in
+        // reclaim mode, this is the final answer
+        vlog(rpc::rpclog.trace, "home_get_units: fast path got={}", result);
+        co_return result;
+    }
+
+    // We didn't get any units, but there might be some on other cores
+    co_await reclaim_to(allowance, addr, false);
+
+    // This can still fail if there were no reclaimable units, but
+    // we did our best.
+    result = try_get_units(allowance);
+    vlog(
+      rpc::rpclog.trace, "home_get_units: slow (reclaim) path got={}", result);
+    co_return result;
+}
+
+bool conn_quota::try_get_units(home_allowance& allowance) {
+    if (allowance.available) {
+        allowance.available -= 1;
```

Comment:
I don't have a specific scenario in mind, but given the background fibers and all-core operations, one thing that looks at least plausible is some interleaving that temporarily cause values like `allowance.available` to go negative (or in this case very large positive value) and then those intermediate values are sampled, even if overall the value is correct eventually. Do you think that's something to worry about?

Reply:
In this particular location we're implicitly clamping it to a minimum of zero with the `if (allowance.available` .  Similarly in `home_allowance::put` it is clamped to the maximum limit.  The clamping is primarily there to deal with situations where we adjust the limits while some units are in flight, but also helps to avoid risk of things like wrapping in case of bugs.